### PR TITLE
Replace Ace Editor with Monaco Editor

### LINKED ISSUE #431: Switch code editor to Monaco
Switch to Monaco

We are planning to update the code editor to use the [Monaco](https://microsoft.github.io/monaco-editor/) that powers [Visual Studio Code](https://github.com/microsoft/vscode). With this, we will gain an exceptional amount of features and control over the experience of editing code within Winter CMS, using an interface that users are comfortable with.

We will be able to take advantage of the extension work done in our [official VSCode extension](https://github.com/wintercms/vscode-extension/issues/9) to provide the same level of functionality in the Monaco editor.

### PR DESCRIPTION
**Pull request by bennothommo**
Fixes #431. Related: https://github.com/wintercms/wn-builder-plugin/pull/73
Screenshot 2026-01-02 at 1 02 11 AM
Screenshot 2026-01-02 at 1 02 40 AM
Screenshot 2026-01-02 at 1 03 25 AM

This PR introduces the Monaco editor in replacement of the Ace Editor. At this stage for the PR, we are aiming for parity with the functionality previously provided by Ace, with the intention of increasing the functionality after this PR is merged.

### Still to do:

- [x] Support all prior configuration options.
- [x] Re-implement all previous themes, or find suitable alternatives (see https://github.com/brijeshb42/monaco-themes)
- [x] Ensure code editor works in all current core locations (settings areas, CMS editor, Builder plugin)
- [x] Try and figure out how to remove the automated loading of the Codicon font, as it's being implemented in the widget's CSS assets.

### Changes:
- Auto-completing tags preference is no longer available, as Monaco doesn't have this as an option.
- Autocompletion and code snippets preferences are no longer available, as these are more contextual and will be implemented based on what is being edited in the editor.
- Code folding is now a simple boolean.
- The **Crimson Editor** theme has been dropped - I couldn't find a complete replica of this theme, and it looks too similar to other themes in the list.

## Summary by CodeRabbit

* **New Features**
  * Replaced legacy editor with a modern Monaco-based editor: minimap, bracket colors, inline color previews, richer language services, and many new themes.

* **Improvements**
  * Unified editor preferences into a single backend plugin and refreshed editor styling, toolbar, and tab collapse icon behavior.
  * Asset/loading initialization updated for improved editor packaging.

* **Bug Fixes**
  * Fixed duplicated render invocations causing double render handling.

* **Documentation**
  * Added a README for the new editor integration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: modules/backend/ServiceProvider.php [hunk 1]
```diff
@@ -104,6 +105,10 @@ protected function registerAssetBundles()
                 $combiner->registerBundle('~/modules/backend/formwidgets/richeditor/assets/js/build.js');
             }
         });
+
+        MixAssets::instance()->registerCallback(function ($mix) {
+            $mix->registerPackage('module-backend.codeeditor', '~/modules/backend/formwidgets/codeeditor/assets/winter.mix.js');
```
Comment (LukeTowers) [inline modules/backend/ServiceProvider.php]:
```suggestion
            $mix->registerPackage('formwidget.codeeditor', '~/modules/backend/formwidgets/codeeditor/assets/winter.mix.js');
```

File: modules/backend/assets/ui/js/ui/WidgetHandler.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    onReady() {
+        this.initialiseWidgets(document.body);
+    }
+
+    onAjaxUpdate(element) {
+        this.initialiseWidgets(element);
+    }
+
+    initialiseWidgets(element) {
```
Comment (LukeTowers) [inline modules/backend/assets/ui/js/ui/WidgetHandler.js]: We should stick to the US spelling of initialize: https://github.com/search?q=org%3Awintercms+initialize&type=code vs https://github.com/search?q=org%3Awintercms+initialise&type=code

File: modules/backend/controllers/preferences/_field_editor_preview.php [hunk 3]
```diff
@@ -1,22 +1,25 @@
...
-      data-show-print-margin="<?= $model->editor_show_print_margin ?>"
+      data-display-indent-guides="<?= $model->editor_display_indent_guides ? 'true' : 'false' ?>"
+      data-show-print-margin="<?= $model->editor_show_print_margin ? 'true' : 'false' ?>"
       data-show-gutter="<?= $model->editor_show_gutter ? 'true' : 'false' ?>"
+      data-show-minimap="<?= $model->editor_show_minimap ? 'true' : 'false' ?>"
       data-language="css"
       data-margin="0"
-      data-vendor-path="<?= Url::asset('/modules/backend/formwidgets/codeeditor/assets/vendor/ace') ?>/">
-      <textarea name="editorpreferences_codeeditor"><?= e($this->makePartial('example_code')) ?></textarea>
+      data-vendor-path="<?= Url::asset('/modules/backend/formwidgets/codeeditor/assets/vendor/ace') ?>/"
```
Comment (LukeTowers) [inline modules/backend/controllers/preferences/_field_editor_preview.php]: Is this line still required?

**Review by LukeTowers — DISMISSED** [review]

File: modules/backend/assets/ui/js/ui/WidgetHandler.js (hunk 2 above)
Comment (bennothommo) [reply modules/backend/assets/ui/js/ui/WidgetHandler.js]: True - I'll fix the references in the other repos when I can, but I've completed it for this PR.

File: modules/backend/ServiceProvider.php (hunk 1 above)
Comment (bennothommo) [reply modules/backend/ServiceProvider.php]: The reason I named it that way is that I see the code editor as a "subset" of the Backend module. I was thinking of introducing a way to namespace packages (like this) and add an argument to `mix:compile` to compile everything within a namespace.

Comment (LukeTowers) [reply modules/backend/ServiceProvider.php]: Sure, that sounds fine. Could we do `module.backend.formwidgets.codeeditor` perhaps?

Comment (bennothommo) [reply modules/backend/ServiceProvider.php]: It'd be `module-backend.formwidgets.codeeditor`, but yep, that works for me.

**Comment by LukeTowers** [comment]
@bennothommo what's the status of this?

**Comment by bennothommo** [comment]
When I last worked on this, I was having some dramas with the Markdown editor, since that also had to be converted to Monaco. I was hesitant to include another editor just for the Markdown editor, but if we can find a lean one, I'm happy to use that instead and keep Monaco for just the code editing (which it works amazingly with).

**Comment by LukeTowers** [comment]
What's the problem with the markdown editor? Does it currently use ace?

**Comment by bennothommo** [comment]
Yeah it uses Ace currently. I feel keeping Ace around and introducing Monaco for the code editor might be too much "bloat".

**Comment by LukeTowers** [comment]
I'm fine with having both if we can clearly identify all the locations we can remove Ace from as soon as we get a replacement for the usage of Ace in the markdown editor; that way at least it won't hold us up from getting the benefits of monaco and it'll be simple enough to remove Ace when we replace the markdown editor.

Any thoughts on that?

File: modules/backend/formwidgets/codeeditor/assets/js/codeeditor.old.js
Comment (LukeTowers) [inline modules/backend/formwidgets/codeeditor/assets/js/codeeditor.old.js:1]: Is this still required?

File: modules/backend/formwidgets/codeeditor/assets/less/codeeditor.less [hunk 4]
```diff
@@ -1,103 +1,665 @@
...
-        .editor-toolbar {
-            z-index: @zindex-fullscreen + 2;
-        }
-        .editor-code { .border-radius(0); }
-        .ace_search { z-index: @zindex-fullscreen + 3; }
-    }
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
```
Comment (LukeTowers) [inline modules/backend/formwidgets/codeeditor/assets/less/codeeditor.less:90]: Can we import this from a separate file?

File: modules/backend/models/Preference.php [hunk 5]
```diff
@@ -59,6 +59,8 @@ public function initSettingsData()
         $this->editor_font_size = $config->get('editor.font_size', 12);
         $this->editor_word_wrap = $config->get('editor.word_wrap', 'fluid');
         $this->editor_code_folding = $config->get('editor.code_folding', 'manual');
+        // @deprecated v1.2.2
```
Comment (LukeTowers) [inline modules/backend/models/Preference.php]:
```suggestion
        // @deprecated v1.3.0
```

File: modules/system/assets/js/snowboard/main/Snowboard.js [hunk 6]
```diff
@@ -336,7 +336,6 @@ export default class Snowboard {
         const listeners = this.listensToEvent(eventName);
         if (listeners.length === 0) {
             this.debug(`No listeners found for global event "${eventName}"`);
-            return true;
```
Comment (LukeTowers) [inline modules/system/assets/js/snowboard/main/Snowboard.js:339]: Is this intentional?

File: modules/system/assets/js/snowboard/main/Snowboard.js [hunk 7]
```diff
@@ -437,7 +436,6 @@ export default class Snowboard {
         const listeners = this.listensToEvent(eventName);
         if (listeners.length === 0) {
             this.debug(`No listeners found for global promise event "${eventName}"`);
-            return Promise.resolve();
```
Comment (LukeTowers) [inline modules/system/assets/js/snowboard/main/Snowboard.js:440]: Is this intentional?

File: modules/system/assets/ui/font/fa-brands-400.ttf
Comment (LukeTowers) [inline modules/system/assets/ui/font/fa-brands-400.ttf:1]: Why are these files being added?

File: themes/demo/pages/ajax.htm
Comment (LukeTowers) [inline themes/demo/pages/ajax.htm:1]: Unintentional changes to this file?

File: modules/backend/formwidgets/codeeditor/assets/js/codeeditor.old.js
Comment (bennothommo) [reply modules/backend/formwidgets/codeeditor/assets/js/codeeditor.old.js:1]: I'll need it probably up until the last moment, just in case there's any other original functionality I need to port across. I'll remove it once we're ready to go.

File: modules/backend/formwidgets/codeeditor/assets/less/codeeditor.less (hunk 4 above)
Comment (bennothommo) [reply modules/backend/formwidgets/codeeditor/assets/less/codeeditor.less:90]: I can have another look, but from what I remember, the CSS is inlined in Monaco so it may not be importable.

File: modules/system/assets/js/snowboard/main/Snowboard.js (hunk 6 above)
Comment (bennothommo) [reply modules/system/assets/js/snowboard/main/Snowboard.js:339]: Yes, it short-circuits the event listeners if there are none available.

File: modules/system/assets/js/snowboard/main/Snowboard.js (hunk 7 above)
Comment (bennothommo) [reply modules/system/assets/js/snowboard/main/Snowboard.js:440]: Yes, it short-circuits the event listeners if there are none available.

File: modules/system/assets/ui/font/fa-brands-400.ttf
Comment (bennothommo) [reply modules/system/assets/ui/font/fa-brands-400.ttf:1]: Possibly done before we changed how Font Awesome was being built. I'll remove.

File: themes/demo/pages/ajax.htm
Comment (bennothommo) [reply themes/demo/pages/ajax.htm:1]: Most likely - I'll revert them.

**Comment by LukeTowers** [comment]
@coderabbitai review

File: modules/backend/formwidgets/codeeditor/assets/js/codeeditor.js [hunk 8]
```diff
@@ -1,468 +1,1531 @@
...
+                monaco.languages.registerCodeLensProvider(language, {
+                    provideCodeLenses: (model, token) => {
+                        return provider(model, token);
+                    },
+                    resolveCodeLens: (model, codeLens, token) => {
+                        return resolver(model, codeLens, token) ?? codeLens;
+                    },
+                })
+            );
+        }
```
[the remark this answers is not in the document]
Comment (LukeTowers) [reply modules/backend/formwidgets/codeeditor/assets/js/codeeditor.js:1526]: Callers need to implement the command. Not fixing this for now.

**Review by LukeTowers — APPROVED** [review]
